### PR TITLE
ModAFNO for interpolation

### PR DESCRIPTION
**Pull request by jleinonen**
# Modulus Pull Request

## Description
This adds a Modulated AFNO (ModAFNO) layer that adds a modulation input to the AFNO architecture. This can be used to build tunable AFNO models and works, for example, to build a weather interpolation model.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

## Dependencies
No new dependencies needed

### CODE REVIEW

File: modulus/models/afno/modafno.py
Comment (NickGeneva) [inline modulus/models/afno/modafno.py:73]: Why a plus 1 here?

File: modulus/models/afno/modafno.py
Comment (NickGeneva) [inline modulus/models/afno/modafno.py:167]: type `Literal["complex", "real"]` seems more appropriate here

File: modulus/models/afno/modafno.py
Comment (NickGeneva) [inline modulus/models/afno/modafno.py:257]: In the constructor there is presently no check if `scale_shift_mode` is real or complex.

I would suggest either line 178 gets changed to through an error if not one of these, or line 260 gets changed to an else for consistency.

File: modulus/models/afno/modafno.py
Comment (NickGeneva) [inline modulus/models/afno/modafno.py:121]: Is this ever a variable dim or do we always know the shape will be something like (B, 1, 1, N)

File: modulus/models/afno/modafno.py
Comment (NickGeneva) [inline modulus/models/afno/modafno.py:190]: May consider adding an assert check here to check consistency between x and mod_embed

File: modulus/models/afno/modafno.py
Comment (NickGeneva) [inline modulus/models/afno/modafno.py:384]: Would users realistically want to shut this off? Would this just become AFNO with out these? I wonder if this option is really needed or can just be considered part of the arch.

File: modulus/models/afno/modafno.py
Comment (NickGeneva) [inline modulus/models/afno/modafno.py:424]: Could so some thing like

filder_model = AFNOMlp()
self.filer = lambda x, mode_embed : filder_model(x)

to avoid having an if statement in your forward.

File: modulus/models/afno/modafno.py
Comment (NickGeneva) [inline modulus/models/afno/modafno.py:431]: Similarly, not a big fan of conditionals in the forward.

File: modulus/models/afno/modafno.py
Comment (NickGeneva) [inline modulus/models/afno/modafno.py:637]: maybe assert check should be here

File: modulus/models/afno/modafno.py
Comment (NickGeneva) [inline modulus/models/afno/modafno.py:627]: Just Confirm that dim=1 is always valid and that mod will never be higher dimensional needing something like dim=-1c

File: modulus/models/afno/modafno.py
Comment (NickGeneva) [inline modulus/models/afno/modafno.py:482]: Nit: add `, by default True` and order these the same as the parameters of the constructor.

File: modulus/models/afno/modembed.py
Comment (NickGeneva) [inline modulus/models/afno/modembed.py:42]: -> Tensor

File: modulus/models/afno/modembed.py
Comment (NickGeneva) [inline modulus/models/afno/modembed.py:65]: Nit: return type hint -> Tensor

File: modulus/models/afno/modembed.py
Comment (NickGeneva) [inline modulus/models/afno/modembed.py:111]: Nit: return type hint -> Tensor

File: modulus/models/afno/modembed.py
Comment (NickGeneva) [inline modulus/models/afno/modembed.py:100]: Can you make this and the following line `self.emb = ...` and get rid of the conditional in the forward?

File: modulus/models/afno/modembed.py
Comment (NickGeneva) [inline modulus/models/afno/modembed.py:102]: Nit: add
```
else:
    raise ValueError(f"Embedding {method} not supported")
```

**Comment by NickGeneva** [comment]
Can you add the model to the docs rst file:

https://github.com/NVIDIA/modulus/blob/main/docs/api/modulus.models.rst

File: modulus/models/afno/modafno.py
Comment (jleinonen) [reply modulus/models/afno/modafno.py:73]: The net gives pretty small values after initialization so with the +1 we get scale close to 1 and shift close to 0 in the beginning of training. This turned out to lead to more stable training.

File: modulus/models/afno/modafno.py
Comment (jleinonen) [reply modulus/models/afno/modafno.py:167]: Great, will update - my knowledge of type annotations is rather incomplete.

File: modulus/models/afno/modembed.py
Comment (jleinonen) [reply modulus/models/afno/modembed.py:100]: Good idea, but wouldn't this break compatibility with existing checkpoints (since AFAIK layer names need to match those in the checkpoint file)?

File: modulus/models/afno/modembed.py
Comment (jleinonen) [reply modulus/models/afno/modembed.py]: Fixed in https://github.com/NVIDIA/modulus/pull/547/commits/f634dfdcaae1cc73e0e67568843c7fd7fd5d1d09.

File: modulus/models/afno/modembed.py
Comment (jleinonen) [reply modulus/models/afno/modembed.py]: Fixed in https://github.com/NVIDIA/modulus/pull/547/commits/f634dfdcaae1cc73e0e67568843c7fd7fd5d1d09.

File: modulus/models/afno/modembed.py
Comment (jleinonen) [reply modulus/models/afno/modembed.py]: Fixed in https://github.com/NVIDIA/modulus/pull/547/commits/f634dfdcaae1cc73e0e67568843c7fd7fd5d1d09.

File: modulus/models/afno/modembed.py
Comment (jleinonen) [reply modulus/models/afno/modembed.py:102]: Added in https://github.com/NVIDIA/modulus/pull/547/commits/0b75c9b0348c95a3096ead5437d2222dc18f4418.

File: modulus/models/afno/modafno.py
Comment (jleinonen) [reply modulus/models/afno/modafno.py]: Fixed in https://github.com/NVIDIA/modulus/pull/547/commits/0b75c9b0348c95a3096ead5437d2222dc18f4418.

File: modulus/models/afno/modafno.py
Comment (jleinonen) [reply modulus/models/afno/modafno.py]: Done in https://github.com/NVIDIA/modulus/pull/547/commits/8e24b676645436ffe014e2a584a4c2404f429ec5

File: modulus/models/afno/modafno.py
Comment (jleinonen) [reply modulus/models/afno/modafno.py:190]: This would be an assert for the tensor channel dimension? In any case, I tried to add it and ruff complains:
```modulus/models/afno/modafno.py:191:9: S101 Use of `assert` detected```

File: modulus/models/afno/modafno.py
Comment (jleinonen) [reply modulus/models/afno/modafno.py:121]: Maybe what I had in mind when I wrote this was that (Mod)AFNO can be extended to 1D/3D quite easily and then this would still work... but for the current implementation `x.ndim ==4` always.

File: modulus/models/afno/modafno.py
Comment (jleinonen) [reply modulus/models/afno/modafno.py:639]: This would have the same issue as https://github.com/NVIDIA/modulus/pull/547#discussion_r1646497439?

**Comment by jleinonen** [comment]
> Can you add the model to the docs rst file:
> 
> https://github.com/NVIDIA/modulus/blob/main/docs/api/modulus.models.rst

Added in https://github.com/NVIDIA/modulus/pull/547/commits/5f46f434155c6cf2925eb5e8f0fb4b8cd89f8655, @NickGeneva could you check that this makes sense? Also, shouldn't `AdaptiveFourierNeuralOperator` in that file be just `AFNO`?

File: modulus/models/afno/modafno.py
Comment (jleinonen) [reply modulus/models/afno/modafno.py]: Changed in https://github.com/NVIDIA/modulus/pull/547/commits/447c62e650e79dfc0e4342a93b1bd7780067661d, I implemented this a bit differently to avoid overwriting `self.filter` so existing checkpoints will work. Maybe check if this makes sense to you?

File: modulus/models/afno/modafno.py
Comment (jleinonen) [reply modulus/models/afno/modafno.py]: Changed in https://github.com/NVIDIA/modulus/pull/547/commits/447c62e650e79dfc0e4342a93b1bd7780067661d as with the comment above.

File: modulus/models/afno/modafno.py
Comment (jleinonen) [reply modulus/models/afno/modafno.py:259]: Added error check to constructor in https://github.com/NVIDIA/modulus/pull/547/commits/4750b4d689943825b371709428cd1ae51289de59.

File: modulus/models/afno/modafno.py
Comment (jleinonen) [reply modulus/models/afno/modafno.py:635]: Yes, here `dim==1` should be always valid. The shape of `x` here is `(B, P, C)` where the `P` contains all the AFNO patches flattened into one dimension. So we create a singleton dimension in `mod_additive` to make its shape `(B, 1, C)` to make all patches get the same modulation. The flattening to a `(B, P, C)` shape should always happen (even if this happens to be turned in to a 3D ModAFNO later).

File: modulus/models/afno/modafno.py
Comment (jleinonen) [reply modulus/models/afno/modafno.py:386]: When developing, I was testing this with either `modulate_mlp` and `modulate_filter` turned off. It turns out to work best with both, but it was surprisingly good with only `modulate_mlp` active. Anyway, this should be handled without conditionals in the forward now.

File: modulus/models/afno/modafno.py
Comment (NickGeneva) [reply modulus/models/afno/modafno.py:121]: That makes sense. Was just tracing dims from the model.

File: modulus/models/afno/modafno.py
Comment (NickGeneva) [reply modulus/models/afno/modafno.py:192]: Ah yeah I forgot about the asset checks, ruff will want you to have a if statement and then raise the error yourself.

I suppose the idea is to make sure that the user's x and mod_embed are of the correct sizes and give a easy to understand error if it isnt. Could help make it easier to use for people. Just my opinion though,Thoughts?

File: modulus/models/afno/modembed.py
Comment (NickGeneva) [reply modulus/models/afno/modembed.py:100]: Yeah this would require some manual stitching of the previous checkpoint.

I'd be for making the update, I think the fix for the checkpoint should be a rename of the weight key.
```python
checkpoint = torch.load(load_name)
for key, value in checkpoint['model'].items():
    if key == "modafno.something.something.onehot_embed":
           checkpoint['model']['"modafno.something.something.embed"] = value
```

I think this would be alright (I had to do this before for I think the original AFNO checkpoint :P ). But naturally we need to make sure we have an original implementation to make sure we outputs are exact. Can help if needed.

File: modulus/models/afno/modafno.py
Comment (NickGeneva) [reply modulus/models/afno/modafno.py:73]: Makes sense, just curious :+1:

File: docs/api/modulus.models.rst
Comment (NickGeneva) [inline docs/api/modulus.models.rst:41]: Yes, please update this to AFNO

**Comment by NickGeneva** [comment]
Over all I think edits look good, confirm that the outputs are consistent with existing checkpoint and I think we can merge

**Comment by NickGeneva** [comment]
Oh also please add a bullet in the change log about adding this arch.

**Review by NickGeneva — APPROVED** [review]
Approving to unblock

File: docs/api/modulus.models.rst
Comment (jleinonen) [reply docs/api/modulus.models.rst]: Fixed in [f849071](https://github.com/NVIDIA/modulus/pull/547/commits/f849071daffda7c4c8c055e007e59c42d9c21ed5).

**Comment by jleinonen** [comment]
@NickGeneva I have confirmed that the original checkpoints loads fine into the Modulus version of ModAFNO in this PR, no changes required.

**Comment by jleinonen** [comment]
@NickGeneva could you also nominate another reviewer, we'll need 2 approvals to merge.

**Review by ktangsali — APPROVED** [review]
Looks good! Thanks for also adding the appropriate changes to the docs - appreciate that!

**Pull request merged**
